iptables: Address a TODO, check existing chain before chain deletion/creation

masap opened the pull request:
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
iptables: Address a TODO, check existing chain before chain deletion/creation

The TODO was introduced by commit e045c6ce2fd81b85637894a1655a90bc367b90f7 "Split portals into host and container".

/cc @thockin

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
```docs
NONE
```

masap commented:
@thockin  Is there any comment ?

thockin commented on pkg/util/iptables/iptables.go:
```diff
@@ -246,13 +246,16 @@ func (runner *runner) EnsureChain(table Table, chain Chain) (bool, error) {
+
 	out, err := runner.run(opCreateChain, fullArgs)
 	if err != nil {
-		if ee, ok := err.(utilexec.ExitError); ok {
```
I don't think we can remove this - it's always going to be a race

thockin reviewed (commented):
Hi!  I appreciate the PR, but I don't think we should do this.  Here's why:

We trade one op (try create/delete) for best-case one op (check) and worst case two ops.  Unless we can show that the list op is significantly faster than the try-create/delete op, this is a net negative.  My (very quick!) testing shows that running a list op on an existing chain is about the same cost as just trying to create it. 

I think the right answer is just to remove the TODO and maybe add a comments that it's not worth checking explicitly because a "check" isn't any cheaper than a "do".

masap committed:
iptables: Remove invalid TODO

By measurement, pre-check with "iptables -S" is not faster than existing code
(just deleting a chain). So remove this TODO and add a comment to the place.
Add a similar comment to EnsureChain as well.

masap commented:
> My (very quick!) testing shows that running a list op on an existing chain is about the same cost as just trying to create it.

Thank you for a comment.
Right. My PR is slower or even than previous for both ```EnsureChain``` and ```DeleteChain```.
These are averages of each 1000 times trials (unit is micro second).

||EnsureChain(create)|EnsureChain(not create)|DeleteChain(delete)|DeleteChain(not delete)|
|---|---|---|---|---|
|before|1897|1743|1783|1558|
|after|3735|1708|3627|1761|

> I think the right answer is just to remove the TODO and maybe add a comments
> that it's not worth checking explicitly because a "check" isn't any cheaper than a "do".

I fixed my commit as you said.

masap commented:
@thockin Hi! How about this ?

masap commented:
/ok-to-test

masap commented:
/test pull-kubernetes-unit

masap commented:
/test pull-kubernetes-node-e2e-containerd
